File: services/python/starlette/src/routes/handlers/internal.py
import structlog
from starlette.requests import Request
from starlette.responses import Response

from routes.errors import AppErrors, send_error


async def internal_handler(request: Request, exception: Exception) -> Response:
    structlog.get_logger().error(
        "Internal server error",
        method=request.method,
        path=request.url.path,
        error=repr(exception),
    )

    return send_error(AppErrors.Internal)

File: services/go/echo/routes/middleware/internal.go
package middleware

import (
	"echo/routes"
	"log/slog"

	"github.com/labstack/echo/v5"
)

func RecoverMiddleware(next echo.HandlerFunc) echo.HandlerFunc {
	return func(c *echo.Context) (err error) {
		defer recoverPanic(c, &err)
		return next(c)
	}
}

func recoverPanic(c *echo.Context, err *error) {
	r := recover()
	if r == nil {
		return
	}
	slog.Error("Internal server error", "method", c.Request().Method, "path", c.Request().URL.Path, "error", r)
	*err = routes.AppErrors.Internal.Error(c)
}
